packages/grid/x-data-grid/src/utils/fastShallowCompare.ts

```diff
@@ -0,0 +1,30 @@
+const is = Object.is;
+
+export function fastShallowCompare<T extends Record<string, unknown> | null | undefined>(
+  a: T,
+  b: T,
+) {
+  if (a === b) {
+    return true;
+  }
+  if (!(a instanceof Object) || !(b instanceof Object)) {
+    return false;
+  }
+
+  for (const key in a) {
+    if (!(key in b)) {
+      return false;
+    }
+    if (!is(a[key], b[key])) {
+      return false;
+    }
+  }
+
+  for (const key in b) {
+    if (!(key in a)) {
+      return false;
+    }
+  }
+
+  return true;
+}
```

Comment:
When the store updates, it publishes an event to all `useGridSelector` listeners. For the grid cells, one of those selectors is `.getCellParams()` which returns a bundle of props related to the cell. This function allows us to compare those bundles very quickly and avoid re-renders if none of those fields has changed.

I've looked at a few comparison packages & functions, they all use `Object.keys` twice, and/or `Object.hasOwnProperty`. Using a `for ... in` and `key in ...` allows us to avoid 2 array allocations & many function calls, which ensure this is as fast & resource-efficient as possible. The trade-off is that this function has subtle differences if it receives an exotic object (e.g. something with a prototype), but considering this is an internal function it feels ok to focus on speed.

With the current implementation, store updates don't even show up in the profiling logs because they're too fast.

Reply:
Interesting thought, but I think we could make this run faster by create a comparison function generator, something like `function createShallowCompare() { return fastShallowCompare() {...} }`, and each object type would have its own comparison function, e.g. `const cellParamsCompare = createShallowCompare()`. This would allow JS engines' inline caches to only see specific object shapes go into each of those comparison functions, which could potentially allow them to JIT them to very specialized functions. All speculative though, and this is fine the way it is.